File: backend/routes/notifications.js
// routes/notifications.js
// ─────────────────────────────────────────────────────────
// GET  /api/notifications       → إشعارات المستخدم الحالي
// PUT  /api/notifications/read  → تعليم الكل كمقروء
// ─────────────────────────────────────────────────────────

const express = require('express');
const router = express.Router();
const pool = require('../config/db');
const verifyToken = require('../middleware/auth');

// ── GET /api/notifications ────────────────────────────────
// يجلب الإشعارات الموجّهة للمستخدم أو لدوره
router.get('/', verifyToken, async (req, res) => {
  try {
    const result = await pool.query(
      `SELECT * FROM notifications
       WHERE user_id = $1 OR role = $2
       ORDER BY created_at DESC
       LIMIT 50`,
      [req.user.id, req.user.role]
    );
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'خطأ في السيرفر' });
  }
});

// ── PUT /api/notifications/read ───────────────────────────
// تعليم كل إشعارات المستخدم كمقروءة
router.put('/read', verifyToken, async (req, res) => {
  try {
    await pool.query(
      `UPDATE notifications SET is_read = TRUE
       WHERE user_id = $1 OR role = $2`,
      [req.user.id, req.user.role]
    );
    res.json({ message: 'تم تعليم الكل مقروءاً' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'خطأ في السيرفر' });
  }
});

module.exports = router;

File: backend-py/routers/notifications.py
from decimal import Decimal
from datetime import date, datetime

from fastapi import APIRouter, Depends, HTTPException

from config.db import get_pool
from middleware.auth import get_current_user

router = APIRouter()


def to_json_value(value):
    if isinstance(value, Decimal):
        return float(value)
    if isinstance(value, (date, datetime)):
        return value.isoformat()
    return value


def row_to_dict(row):
    return {key: to_json_value(row[key]) for key in row.keys()}


# GET /api/notifications/
@router.get("/")
async def get_notifications(user=Depends(get_current_user)):
    pool = await get_pool()

    try:
        rows = await pool.fetch(
            """
            SELECT *
            FROM notifications
            WHERE user_id = $1 OR role = $2
            ORDER BY created_at DESC
            LIMIT 50
            """,
            user.get("id"),
            user.get("role")
        )

        return [row_to_dict(row) for row in rows]

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"خطأ في السيرفر: {str(e)}"
        )


# PUT /api/notifications/read
@router.put("/read")
async def mark_notifications_read(user=Depends(get_current_user)):
    pool = await get_pool()

    try:
        await pool.execute(
            """
            UPDATE notifications
            SET is_read = TRUE
            WHERE user_id = $1 OR role = $2
            """,
            user.get("id"),
            user.get("role")
        )

        return {"message": "تم تعليم الكل مقروءاً"}

    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"خطأ في السيرفر: {str(e)}"
        )